backend/test/unit/openai.test.ts

```diff
@@ -1,63 +1,26 @@
-import { CHAT_MESSAGE_TYPE, ChatHistoryMessage } from '@src/models/chat';
-import { DEFENCE_ID, Defence } from '@src/models/defence';
-import { LEVEL_NAMES } from '@src/models/level';
-import {
-	getValidModelsFromOpenAI,
-	setSystemRoleInChatHistory,
-} from '@src/openai';
+import { beforeEach, describe, expect, jest, test } from '@jest/globals';
+import { OpenAI } from 'openai';
 
-// Define a mock implementation for the createChatCompletion method
-const mockCreateChatCompletion = jest.fn();
-let mockModelList: { id: string }[] = [];
-jest.mock('openai', () => ({
-	OpenAI: jest.fn().mockImplementation(() => ({
-		chat: {
-			completions: {
-				create: mockCreateChatCompletion,
-			},
-		},
-		models: {
-			list: jest.fn().mockImplementation(() => ({
-				data: mockModelList,
-			})),
-		},
-	})),
-}));
+import { getValidModelsFromOpenAI } from '@src/openai';
 
-jest.mock('@src/openai', () => {
-	const originalModule =
-		jest.requireActual<typeof import('@src/openai')>('@src/openai');
-	return {
-		...originalModule,
-		initOpenAi: jest.fn(),
-		getOpenAI: jest.fn(),
-	};
-});
+jest.mock('openai');
+jest.mock('@src/defence');
 
-jest.mock('@src/langchain', () => {
-	const originalModule =
-		jest.requireActual<typeof import('@src/langchain')>('@src/langchain');
-	return {
-		...originalModule,
-		initQAModel: jest.fn(),
-		initDocumentVectors: jest.fn(),
-	};
-});
+describe('getValidModelsFromOpenAI', () => {
+	const mockOpenAI = OpenAI as jest.MockedClass<typeof OpenAI>;
+	mockOpenAI.prototype = {
+		models: {
+			list: jest.fn(),
+		},
+	} as unknown as typeof mockOpenAI.prototype;
```

Comment:
Don't think this is the correct way to mock a class, you should generally leave prototypes alone. This seems to work:

```
const mockListFn = jest.fn<OpenAI.Models['list']>();
jest.mocked(OpenAI).mockImplementation(() => ({
	models: {
		list: mockListFn,
	},
} as unknown as jest.MockedObject<OpenAI>));
```

Then in your test:

```
mockListFn.mockResolvedValueOnce({
	data: mockModelList,
} as OpenAI.ModelsPage);
```

For the sake of completion, you might want to include an afterEach() to reset mockListFn.

Reply:
yup, seems nicer